File: 540-single-element-in-a-sorted-array/single-element-in-a-sorted-array.cpp
class Solution {
public:
    int singleNonDuplicate(vector<int>& nums) {
        int h = nums.size() - 1;
        int l = 0;
        if (h == 0) return nums[0];
        else if(nums[0] != nums[1]) return nums[0];
        else if(nums[h] != nums[h - 1]) return nums[h];
        while(h >= l){
        int mid = l + (h - l) / 2;
        if (nums[mid] != nums[mid - 1] && nums[mid] != nums[mid + 1]) return nums[mid];
        if (mid % 2 == 1 && nums[mid] == nums[mid - 1] || mid % 2 == 0 && nums[mid] == nums[mid + 1]) l = mid + 1;// eliminating the half as no duplicates found
        else {
            h = mid - 1;
        }
        }
        return -1;
        
    }
};

File: 540-single-element-in-a-sorted-array/single-element-in-a-sorted-array.py
class Solution:
    def isToLeft(self, mid: int, nums: List[int]) -> bool:
        if mid == len(nums) - 1:
            return True
        elif (mid % 2):#if its 0 -> false means this is true for odd
            return(nums[mid]!= nums[mid - 1])
        else: 
            return(nums[mid]!= nums[mid + 1])

    def singleNonDuplicate(self, nums: List[int]) -> int:
        l = 0
        r = len(nums) - 1
        while (l <= r) :
            mid = (l+r) // 2
            if self.isToLeft(mid, nums) : 
                ans = nums[mid]
                r = mid - 1
            else : l = mid + 1
             
        return ans